File: binding/src/libqsharedmemory.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsharedmemory.html
pub const qsharedmemory = struct {

    /// New constructs a new QSharedMemory object.
    pub fn New() QtC.QSharedMemory {
        return qtc.QSharedMemory_new();
    }


    /// New2 constructs a new QSharedMemory object.
    pub fn New2(key: ?*anyopaque) QtC.QSharedMemory {
        return qtc.QSharedMemory_new2(@ptrCast(key));
    }


    /// New3 constructs a new QSharedMemory object.
    pub fn New3(key: []const u8) QtC.QSharedMemory {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_new3(key_str);
    }


    /// New4 constructs a new QSharedMemory object.
    pub fn New4(parent: ?*anyopaque) QtC.QSharedMemory {
        return qtc.QSharedMemory_new4(@ptrCast(parent));
    }


    /// New5 constructs a new QSharedMemory object.
    pub fn New5(key: ?*anyopaque, parent: ?*anyopaque) QtC.QSharedMemory {
        return qtc.QSharedMemory_new5(@ptrCast(key), @ptrCast(parent));
    }


    /// New6 constructs a new QSharedMemory object.
    pub fn New6(key: []const u8, parent: ?*anyopaque) QtC.QSharedMemory {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_new6(key_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetKey(self: ?*anyopaque, key: []const u8) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSharedMemory_SetKey(@ptrCast(self), key_str);
    }

    pub fn Key(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_Key(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.Key: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetNativeKey(self: ?*anyopaque, key: ?*anyopaque) void {
        qtc.QSharedMemory_SetNativeKey(@ptrCast(self), @ptrCast(key));
    }

    pub fn SetNativeKey2(self: ?*anyopaque, key: []const u8) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSharedMemory_SetNativeKey2(@ptrCast(self), key_str);
    }

    pub fn NativeKey(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_NativeKey(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.NativeKey: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn NativeIpcKey(self: ?*anyopaque, ) QtC.QNativeIpcKey {
        return qtc.QSharedMemory_NativeIpcKey(@ptrCast(self));
    }

    pub fn Create(self: ?*anyopaque, size: i64) bool {
        return qtc.QSharedMemory_Create(@ptrCast(self), size);
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QSharedMemory_Size(@ptrCast(self));
    }

    pub fn Attach(self: ?*anyopaque, ) bool {
        return qtc.QSharedMemory_Attach(@ptrCast(self));
    }

    pub fn IsAttached(self: ?*anyopaque, ) bool {
        return qtc.QSharedMemory_IsAttached(@ptrCast(self));
    }

    pub fn Detach(self: ?*anyopaque, ) bool {
        return qtc.QSharedMemory_Detach(@ptrCast(self));
    }

    pub fn Data(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSharedMemory_Data(@ptrCast(self));
    }

    pub fn ConstData(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSharedMemory_ConstData(@ptrCast(self));
    }

    pub fn Data2(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QSharedMemory_Data2(@ptrCast(self));
    }

    pub fn Lock(self: ?*anyopaque, ) bool {
        return qtc.QSharedMemory_Lock(@ptrCast(self));
    }

    pub fn Unlock(self: ?*anyopaque, ) bool {
        return qtc.QSharedMemory_Unlock(@ptrCast(self));
    }

    pub fn Error(self: ?*anyopaque, ) i32 {
        return qtc.QSharedMemory_Error(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsKeyTypeSupported(typeVal: u16) bool {
        return qtc.QSharedMemory_IsKeyTypeSupported(@intCast(typeVal));
    }

    pub fn PlatformSafeKey(key: []const u8) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_PlatformSafeKey(key_str);
    }

    pub fn LegacyNativeKey(key: []const u8) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_LegacyNativeKey(key_str);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSharedMemory_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsharedmemory.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetNativeKey22(self: ?*anyopaque, key: []const u8, typeVal: u16) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSharedMemory_SetNativeKey22(@ptrCast(self), key_str, @intCast(typeVal));
    }

    pub fn Create2(self: ?*anyopaque, size: i64, mode: i32) bool {
        return qtc.QSharedMemory_Create2(@ptrCast(self), size, @intCast(mode));
    }

    pub fn Attach1(self: ?*anyopaque, mode: i32) bool {
        return qtc.QSharedMemory_Attach1(@ptrCast(self), @intCast(mode));
    }

    pub fn PlatformSafeKey2(key: []const u8, typeVal: u16) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_PlatformSafeKey2(key_str, @intCast(typeVal));
    }

    pub fn LegacyNativeKey2(key: []const u8, typeVal: u16) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSharedMemory_LegacyNativeKey2(key_str, @intCast(typeVal));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSharedMemory_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/sharedmemory.html#types
pub const enums = struct {
    pub const AccessMode = enum {
        pub const ReadOnly: i32 = 0;
        pub const ReadWrite: i32 = 1;
    };

    pub const SharedMemoryError = enum {
        pub const NoError: i32 = 0;
        pub const PermissionDenied: i32 = 1;
        pub const InvalidSize: i32 = 2;
        pub const KeyError: i32 = 3;
        pub const AlreadyExists: i32 = 4;
        pub const NotFound: i32 = 5;
        pub const LockError: i32 = 6;
        pub const OutOfResources: i32 = 7;
        pub const UnknownError: i32 = 8;
    };

};

File: binding/src/libqsharedmemory.cpp
#include <QNativeIpcKey>
#include <QObject>
#include <QSharedMemory>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qsharedmemory.h>
#include "libqsharedmemory.h"
#include "libqsharedmemory.hxx"

QSharedMemory* QSharedMemory_new() {
	 return new QSharedMemory();
}

QSharedMemory* QSharedMemory_new2(const QNativeIpcKey* key) {
	 return new QSharedMemory(*key);
}

QSharedMemory* QSharedMemory_new3(const libqt_string key) {
	 return new QSharedMemory(QString::fromUtf8(key.data, key.len));
}

QSharedMemory* QSharedMemory_new4(QObject* parent) {
	 return new QSharedMemory(parent);
}

QSharedMemory* QSharedMemory_new5(const QNativeIpcKey* key, QObject* parent) {
	 return new QSharedMemory(*key, parent);
}

QSharedMemory* QSharedMemory_new6(const libqt_string key, QObject* parent) {
	 return new QSharedMemory(QString::fromUtf8(key.data, key.len), parent);
}

libqt_string QSharedMemory_Tr(const char* s) {
	QString _ret = QSharedMemory::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSharedMemory_SetKey(QSharedMemory* self, const libqt_string key) {
	self->setKey(QString::fromUtf8(key.data, key.len));
}

libqt_string QSharedMemory_Key(const QSharedMemory* self) {
	QString _ret = self->key();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSharedMemory_SetNativeKey(QSharedMemory* self, const QNativeIpcKey* key) {
	self->setNativeKey(*key);
}

void QSharedMemory_SetNativeKey2(QSharedMemory* self, const libqt_string key) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len));
}

libqt_string QSharedMemory_NativeKey(const QSharedMemory* self) {
	QString _ret = self->nativeKey();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QNativeIpcKey* QSharedMemory_NativeIpcKey(const QSharedMemory* self) {
	return new QNativeIpcKey(self->nativeIpcKey());
}

bool QSharedMemory_Create(QSharedMemory* self, ptrdiff_t size) {
	return self->create(size);
}

ptrdiff_t QSharedMemory_Size(const QSharedMemory* self) {
	return self->size();
}

bool QSharedMemory_Attach(QSharedMemory* self) {
	return self->attach();
}

bool QSharedMemory_IsAttached(const QSharedMemory* self) {
	return self->isAttached();
}

bool QSharedMemory_Detach(QSharedMemory* self) {
	return self->detach();
}

void* QSharedMemory_Data(QSharedMemory* self) {
	return self->data();
}

const void* QSharedMemory_ConstData(const QSharedMemory* self) {
	return self->constData();
}

const void* QSharedMemory_Data2(const QSharedMemory* self) {
	return self->data();
}

bool QSharedMemory_Lock(QSharedMemory* self) {
	return self->lock();
}

bool QSharedMemory_Unlock(QSharedMemory* self) {
	return self->unlock();
}

int QSharedMemory_Error(const QSharedMemory* self) {
	return self->error();
}

libqt_string QSharedMemory_ErrorString(const QSharedMemory* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QSharedMemory_IsKeyTypeSupported(quint16 typeVal) {
	return QSharedMemory::isKeyTypeSupported(static_cast<QNativeIpcKey::Type>(typeVal));
}

QNativeIpcKey* QSharedMemory_PlatformSafeKey(const libqt_string key) {
	return new QNativeIpcKey(QSharedMemory::platformSafeKey(QString::fromUtf8(key.data, key.len)));
}

QNativeIpcKey* QSharedMemory_LegacyNativeKey(const libqt_string key) {
	return new QNativeIpcKey(QSharedMemory::legacyNativeKey(QString::fromUtf8(key.data, key.len)));
}

libqt_string QSharedMemory_Tr2(const char* s, const char* c) {
	QString _ret = QSharedMemory::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSharedMemory_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSharedMemory::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSharedMemory_SetNativeKey22(QSharedMemory* self, const libqt_string key, quint16 typeVal) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len), static_cast<QNativeIpcKey::Type>(typeVal));
}

bool QSharedMemory_Create2(QSharedMemory* self, ptrdiff_t size, int mode) {
	return self->create(size, static_cast<QSystemSemaphore::AccessMode>(mode));
}

bool QSharedMemory_Attach1(QSharedMemory* self, int mode) {
	return self->attach(static_cast<QSystemSemaphore::AccessMode>(mode));
}

QNativeIpcKey* QSharedMemory_PlatformSafeKey2(const libqt_string key, quint16 typeVal) {
	return new QNativeIpcKey(QSharedMemory::platformSafeKey(QString::fromUtf8(key.data, key.len), static_cast<QNativeIpcKey::Type>(typeVal)));
}

QNativeIpcKey* QSharedMemory_LegacyNativeKey2(const libqt_string key, quint16 typeVal) {
	return new QNativeIpcKey(QSharedMemory::legacyNativeKey(QString::fromUtf8(key.data, key.len), static_cast<QNativeIpcKey::Type>(typeVal)));
}

void QSharedMemory_Delete(QSharedMemory* self) {
    delete self;
}

